mds: new command to flush rstats up to a dir

xxhdx1985126 opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
The first step to make applications like rsync "rstats-aware"

The basic idea is to continously nudge nestlocks under the specified
dir until all those locks and their inode's parents' nestlocks are not
dirty.

A new ceph mds op "CEPH_MDS_OP_RSTATFLUSH" is added, this op
is supposed to be sent out by clients' ioctl function and will trigger
the rstat flush procedure

Please refer to descriptions of commits and comments in the source
code for implementation details.

Fixes: http://tracker.ceph.com/issues/24725
Signed-off-by: Xuehan Xu xuxuehan@360.cn

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test docs`
- `jenkins render docs`

</details>

xxhdx1985126 commented:
@batrick @ukernel Hi, I've refactored the rstat flush PR, please take a look:-) Thanks:-)

xxhdx1985126 commented:
jenkins retest this please

xxhdx1985126 commented:
@batrick @ukernel Hi, until the commit "mds: flush all rstat to root", I refactored the source code as zheng suggested, then I did some more work to add the ability to propagate rstat to some specific directory. After that, I added a ioctl command for the fuse's ioctl function and fixed some bugs. Please take a look:-)

xxhdx1985126 commented:
jenkins retest this please

batrick commented:
Supersedes #22781

ukernel commented on src/mds/ScatterLock.h:
```diff
@@ -114,6 +121,45 @@ class ScatterLock : public SimpleLock {
+
+  void add_nudged_by(metareqid_t rqt) {
+    nudged_by.insert(rqt);
+  }
```
unused

ukernel commented on src/mds/Locker.h:
```diff
@@ -157,10 +159,21 @@ class Locker {
+  // comments in Locker::nude_updated_scatterlocks for why they need
+  // to be traced). A dirty/flushing nestlock is either in
+  // "updated_scatterlocks" or "nudging_nestlocks"
+  xlist<ScatterLock*> nudging_nestlocks;
```
only add item to nudging_nestlocks, but list never get used?

ukernel commented on src/mds/mdstypes.h:
```diff
@@ -1035,6 +1041,17 @@ struct fnode_t {
+  // These two fields won't be stored in the underlying storage system,
+  // so they can be set without pre_dirty
+  utime_t last_rstat_dirty_from;
+  bool rstat_dirty_from_delivered = false;
```
why not set rstat_dirty_from to zero at the same time mds setting pf->accounted_rstat = ipf>rstat

xxhdx1985126 replied:
@ukernel Hi, zheng, thanks for your review

xxhdx1985126 replied:
@ukernel Hi, zheng, thanks for your review

The reason why I added a seperate message as a signal to put rstat_dirty_from to zero is that I thought there could be such cases as:

1. mds.a encodes its lock state and send it to mds.b
2. mds.b fails before it can store the lock state to the underlying storage
3. another rstat whose rstat_dirty_from is later than that was sent to mds.b come up at mds.a
4. if we set rstat_dirty_from to 0 at the time mds setting pf->accounted_rstat to pf_rstat, then when mds.b come up again, it would receive a later rstat_dirty_from, which means the ealier one is lost.

So to make the rstat_dirty_from reliable, I added that RSTATDELIVERED message.

Am I right? Thanks:-)

ukernel replied:
at step 1 and 2, rstat_dirty_from is left untouched. 
a step 3, mds rstat_dirty_from should be unchanged too, because it's value is smaller.

furthermore, rstat is no dirty when accounted_rstat == rstat. what's the point to set rstat_dirty_from

xxhdx1985126 replied:
oh, i think i get your point. do you mean that the accounted_rstat's sync to rstat is protected by the wrlock of nestlock which wouldn't be taken if the inodes' replica is offline, so, when accounted_rstat == rstat, mds.b must have already done with the previous rstat. Right?

If so, i think i should correct that, thank you:)

ukernel commented on src/mds/Locker.cc:
```diff
@@ -4794,12 +4829,113 @@ void Locker::scatter_tick()
+
+          dout(20) << __func__ << " " << mdr->reqid << " nestlock of " << *lock->get_parent()
+            << " already nudged, adding callback" << dendl;
+          lock->add_waiter(SimpleLock::WAIT_STABLE, new C_Locker_Nudge(this, mdr));
```
here allocate a C_Locker_Nudge but not break the loop.   The loop may create lots of C_Locker_Nudge, it's very inefficiency.

I prefer to use another list to track locks that can be nudged at once. just let scatter_tick to process that list later

xxhdx1985126 replied:
Yes, I'll try to follow your way:-)

xxhdx1985126 replied:
@ukernel At the point you are refering to, only those locks that are of the rstat flush requests' interest will be added a waiter. If a lock is nudged at once, its rstat_dirty_from would be zero, and it wouldn't be interested by rstat flush requests, so they wouldn't cause a C_Locker_Nudge creation:-)

xxhdx1985126 replied:
the CInode::need_to_nudge() is controlling whether a nestlock is interested:-)

ukernel replied:
If the 'N' nestlocks (you are interested in) can not be nudged at the moment, you create 'N'  instance of C_Locker_Nudge.  Each one will iterate whole updated_scatterlock list when it get executed. consider that case that 'N' is large and updated_scatterlock list is also large

xxhdx1985126 replied:
@ukernel I got your poirnt:-) Will fix it:-) Thanks:-)

ukernel commented on src/mds/Locker.cc:
```diff
@@ -5485,3 +5501,17 @@ void Locker::handle_file_lock(ScatterLock *lock, const cref_t<MLock> &m)
+    lock->take_waiting(mask, *finishers);
+  else
+    lock->finish_waiters(mask, r);
+}
```
for the first case in your commit message. the nestlock should already be nudging, I don't see why need to add it to updated_scatterlocks.

for the second case, I think it's easy to make scatter_nudge to re-queue lock to updated_scatterlocks

xxhdx1985126 replied:
Hi, I thought the first case may not always be triggered by a scatter_nudge. it may be triggered by a wrlock_start or something like that, right? So I that it would be safe to uniformly requeue all those nestlocks that just turned stable to updated_scatterlocks.

Am I right?

ukernel replied:
yes, it can happen at xxx_start().    but nestlock's is in LOCK_MIX_FOO state, it will gather dirty rstats.

xxhdx1985126 replied:
I think I understand that, but these two are only the known cases that lead the lost of dirty rstat, there may be others as well. I added this to make sure that dirty rstat are always taken care of, i don't see why it has to be removed

xxhdx1985126 replied:
for the second case, there doesn't be much difference between putting the lock back to updated_scatterlocks, letting the waiter context to scatter_nudge it and scatter_nudging it, letting the waiter context to.check on the lock and wait for it.

besides, these are only cases that i encountered when doing the test, there could be other cases. putting the locks back to updated_scatterlocks can prevent all those cases from happenning once and for all, and even if there isn't other such cases, doing this wouldn't bring anything bad. Right?

xxhdx1985126 replied:
oh, i might have misunderstood you. do you mean that i should simply change the "mark_updated_scatterlock" to a "scatter_nudge"?

ukernel replied:
I'm not fan of the lock_stablized approach. I prefer
1. add lock to updated_scatterlocks when it become dirty
2. if scatter tick failed to nudge a lock, re-queue it to updated_scatterlocks or another list

xxhdx1985126 replied:
yes, i'm already doing that, my approach is like this:
1. add lock to updated_scatterlocks when it become dirty
2. if scatter tick fails to nudge a lock, queue it to nudging_nestlocks

but, there are circumstances that when a lock finished a nudge, it is still dirty and not in the updated_scatterlocks, so we have to put it back. If we don't add a method like lock_stablized, we have add a mark_updated_scatterlocks anywhere there is a "finish_waiters(WAIT_STABLE)" or "take_waiting(WAIT_STABLE)". this would be redundant code, right? Why don't be process the finish_waiter(WAIT_STABLE) at a uniform place? Am I right?

xxhdx1985126 replied:
I think I know what's the problem here.  Although the scatter_nudge's return is a bool value, it's not an indicator whether the lock is nudged immediately or nudge-able.

Actually, even if the lock is nudged immediately, the scatter_nudge will still end up waiting for the lock to be stable. it seems there is currently no way to judge, even in the scatter_nudge method, whether the waiting is caused by an immediate nudge or an lock's scatter-gather process.

The reason why i changed to scatter_nudge's return type to bool is that, at the time, I can't be sure whether the waiter i passed to scatter_nudge will finally be added to some waiting list, if not, i have to delete it to prevent a memory leak, so i changed its return type to bool just to indicate whether the waiter is on some list.

i'm sorry for this miss understanding. i think i can improve that part of the code:). really sorry.

xxhdx1985126 replied:
the scatter_tick is also calling nudge_updated_scatterlocks now

xxhdx1985126 replied:
@ukernel Hi:-)

I think I can change the scatter_mix, simple_lock and simple_sync to return whether the lock is nudged or undergoing a scatter-gather process, so that the scatter_nudge can indicate whether the lock is nudged immediately, just like what I did in the previous submit. 

Then I can make "nudge_updated_scatterlock" to requeue the not-nudged-immediately nestlocks to updated_scatterlocks to prevent dirty rstat nestlock leak. I think this is what you are suggesting, right? Thanks:-)

xxhdx1985126 replied:
@ukernel Or, we may not need to change scatter_mix, simple_lock and simple_sync; we can just make scatter_nudge treat an invocation of scatter_mix, simple_lock or simple_sync as a indication of a successful nudge, as scatter_gather can be seen as part of a immediate nudge. Am I right? Thanks:-)

xxhdx1985126 replied:
@ukernel But I think, although we don't have to put the dirty nestlocks back to updated_scatterlocks when they turn stable, we still have to move them out of the "nudging_nestlocks" queue after they turn stable, so, once agian, a method like lock_stablized seems inevitable, is this right?

xxhdx1985126 replied:
@ukernel Or, I think maybe I can keep locks in updated_scatterlocks until it's not dirty. I mean, instead of poping every lock that needs to be nudged, we keep them in the queue even if they are nudged immediately and let later run of nudge_updated_scatterlocks() to seletively remove not-dirty locks. How do you like this way:-) Thanks:-)

xxhdx1985126 replied:
@ukernel Or, I think maybe we can change the nudging_nestlocks to flushing_nestlocks. And instead of poping every lock out of the updated_scatterlocks queue, we make nudge_updated_scatterlocks() not modify updated_scatterlocks queue, and let the Scatterlock::start_flush() to move a lock from updated_scatterlocks to flushing_nestlocks if that lock's parent doesn't have children with dirty rstat. With this approach, no dirty lock would be lost, since it's only when they "start_flush" and doesn't have any children with dirty rstat does they get to be remove from updated_scatterlocks(), and we can avoid the lock_stablized likens. Am I right about this? Thanks:-)

xxhdx1985126 replied:
@ukernel I think maybe we can remove the nudging_nestlocks and only rely on the updated_scatterlocks. The approach is like this: instead of poping every lock from the updated_scatterlocks, we make nudge_updated_scatterlock not change the updated_scatterlocks and let the ScatterLock::start_flush to remove the lock from updated_scatterlocks if the lock doesn't have any children with dirty rstat, which is exactly what "not dirty" means. Meanwhile, we make sure every nudge eventually lead to a "start_flush". With this approach, we can avoid the lock_stablized likes and make sure no dirty rstat is lost. Do you think this is feasible? Thanks:-)

xxhdx1985126 replied:
> for the first case in your commit message. the nestlock should already be nudging, I don't see why need to add it to updated_scatterlocks.
> 
> for the second case, I think it's easy to make scatter_nudge to re-queue lock to updated_scatterlocks

Finally, I got your point:-)

xxhdx1985126 replied:
@ukernel Sorry that I didn't catch up with you for so long:-)

ukernel commented on src/mds/ScatterLock.h:
```diff
@@ -45,8 +46,11 @@ class ScatterLock : public SimpleLock {
     DIRTY            = 1 << 10,
     FLUSHING         = 1 << 11,
     FLUSHED          = 1 << 12,
+    NUDGING          = 1 << 13,
```
the 'nudging' state is similar to 'flushing' state.

xxhdx1985126 replied:
Yes, they are similar, but slightly different.

For example, when doing nudge, it may have to wait for the inode to "unfreeze" or "singleauth", during which the lock would still be dirty, not flushing, but we still have to mark it, so that other rstat flush operation can find it if the lock is what they need, too

ukernel commented on src/mds/Locker.cc:
```diff
@@ -4794,12 +4829,113 @@ void Locker::scatter_tick()
+      // can be missed.
+      lock->set_nudging(nudging_nestlocks,
+          (mdr && ino->need_to_nudge(rstat_propagate_time))
+            ? mdr->reqid : metareqid_t());
```
set 'nudging' before checking if nestlock is nudge-able? Is it too early?

xxhdx1985126 replied:
The nudging state's life time is from the point the scatter_nudge is invoked to that the nestlock finish waiting(unfreeze, single auth, or stable). the reason why we have to inclose these waiting interval is that the nestlock isn't in the updated_scatterlocks queue, so we have to trace them.

ukernel replied:
how about
1. add these not-nudge-able lock to another list, and add a waiter context.
2. when waiter context is executed, move it back to updated_scatterlocks list.

with this approach, we don't need the 'nugding'  state

xxhdx1985126 replied:
I don't quite follow you. the nudge events are all independent and of different kinds(unfreeze, stable, single auth), how to let add a waiter for all those events? and what if some other rstat flush op arrives before the waiter is invoked and it has

xxhdx1985126 replied:
I don't quite follow you. the nudge events are all independent and of different kinds(unfreeze, stable, single auth), how to let add a waiter for all those events? and what if some other rstat flush op arrives before the waiter is invoked?  there could be locks in that queue that are of the new rstat flush's interest, too. how to make it wait for those locks?

ukernel replied:
these events can't happen at the same time. just add waiter for individual event. 

rstat flusher just check the if any locks (it interests) is in the new list and wait.

xxhdx1985126 replied:
Oh, finally, I got your point:-)
Actually, I'm already doing exactly what you are saying:-)

The Locker::nudging_nestlocks is the list that you're talking about; the ScatterLock::set_nudging() method is to add the nestlock to the nudging_nestlocks list; ScatterLock::is_nudging() method is to test whether a nestlock is in the nudging_nestlocks list; ScatterLock::clear_nudging() method is to remove a nestlock from the nudging_nestlocks. 

The only difference is that I added a redundant NUDGING state, and this state seems really unnecessary now, all those methods can be easily modified to run without this state. I think I can simply remove the state:-)

ukernel commented on src/mds/Locker.cc:
```diff
@@ -4711,19 +4728,29 @@ void Locker::mark_updated_scatterlock(ScatterLock *lock)
     dout(10) << "scatter_nudge waiting for unfreeze on " << *p << dendl;
-    if (c) {
+    if (c && remote_rstatflush_asker == MDS_RANK_NONE) {
       p->add_waiter(MDSCacheObject::WAIT_UNFREEZE, c);
```
please explain why  'adding a unfreeze waiter' does not work, you need to modify the unfreeze_tree code

xxhdx1985126 replied:
Hi, I thought there could exists chances that when the unfreeze waiter is invoked, the inode that it waited on has already been trimmed out of the memory if, after unfreeze it's not auth anymore, which means the scatter_nudge for that inode's nestlock won't be able to start.

So I added a "asker" to send out a nudge message further to the new auth, or do the nudge if the inode is still auth.

Am I right? Thanks:-)

ukernel replied:
how about adding some code to Migrator::decode_import_inode(). the new auth mds should also contain a MDRequest (for propagating rstat)

xxhdx1985126 replied:
Yes, I'm trying that now:-)

xxhdx1985126 replied:
@ukernel Sorry, I don't quite follow you about this.
The problem is like this, say there is a directory /a/b/c/d/e/f/g/h:

1. /a is on mds.X
2. /a/b/c/d is on mds.Y
3. a rstat flush op want to flush rstat of subtree /a
4. mds.X send a slave request to mds.Y to first flush rstat flush subtree /a/b/c/d, the last step of which is to nudge the inode "/a/b/c/d"
5. mds.Y sends a LOCK_AC_NUDGE to mds.X  to nudge inode of "/a/b/c/d", but at that time subtree /a/b is being migrated to mds.Z, which would lead to the nudge of "/a/b/c/d" suspended.

As you can see, as for now, the processing of LOCK_AC_NUDGE doesn't add any waiter for the unfreeze event, which means when the migration of subtree /a/b finishes, there won't be a nudge of inode "/a/b/c/d", so the slave request that mds.X send out to mds.Y to flush rstat of "/a/b/c/d" will wait on the nudge of inode "/a/b/c/d" infinitely if there is no further action. And even if we add a waiter for the unfreeze event when processing the LOCK_AC_NUDGE message, we may encounter the problem that I mentioned above.

I think if we want to solve this problem by adding code in Migrator::decode_import_inode(), the only solution is to add a waiter that will unconditionally trigger a scatter_nudge of inode "/a/b/c/d"  when the migration finishes in Migrator::decode_import_inode(). Is this what you are suggesting? Thank you:-)

ukernel replied:
For the case above, mds.x will send OP_PROPAGATERSTATS to mds.z. I don't see why it needs special handling.

xxhdx1985126 replied:
@ukernel After mds.X instructed mds.Y to flush its rstats, it may wait on that instruction to return. And at the time mds.X instructed mds.Y, the subtree /a/b is still on mds.X, so it wouldn't send OP_PROPAGATERSTATS. So the situation is mds.X wait on mds.Y, and mds.Y is waiting on a nudge to finish, while, unfortunately, the nudge is lost as the lock it's trying to nudge is being migrated.

xxhdx1985126 replied:
@ukernel If we make sure scatter_nudge requeue lock to updated_scatterlocks whenever they are dirty before the scatter_nudge returns, later scatter_tick can nudge that lock again, which would prevent this forever waiting. Do you think this is an acceptable solution? Thanks:-)

ukernel replied:
I prefer the tick approach.

xxhdx1985126 replied:
@ukernel Ok:-)

xxhdx1985126 replied:
@ukernel Hi, by the way, please take a look at [29721](https://github.com/ceph/ceph/pull/29721) when you are free:-) It's also needed for this PR to run correctly:-)

ukernel commented on src/mds/Locker.cc:
```diff
@@ -935,6 +938,27 @@ void Locker::eval_gather(SimpleLock *lock, bool first, bool *pneed_issue, MDSCon
+            // 5. the previous finish_scatter_update finishes, the codes reaches here and "start_flush"
+            //    the current lock, which would clear the dirty flag of the current lock.
+            // As the CInode::encode_lock_state method doesn't pop dirty_rstat_inodes' rstat to the corresponding
+            // dir, step 5 actually lead to the lost of the dirty state marked by step 4.
```
maybe this is an indication that we should use separate flag for dirty rstat and direcotory has children with dirty rstat

xxhdx1985126 replied:
Yes, we can try that:-)

xxhdx1985126 replied:
@ukernel Hi, it seems that the dirty state is now relied upon by many other function/modules, splitting it would involve a lot of modification. On the other hand, the problem is caused by the clearing of the dirty state not taking into account cases that there are children with dirty rstat, so I think it should be enough to make ScatterLock::clear_dirty to test whether there is such children and not clear the dirty state if it does. Am I right? Thanks:-)

xxhdx1985126 replied:
@ukernel And if the dirty state flag does need to be splitted, i think we can put that in future patches. After all, we can't solve everything within a single PR, am I right? Thanks:-)

xxhdx1985126 replied:
@ukernel Hi, I've seperated the commit for solving this problem into another PR: [29721](https://github.com/ceph/ceph/pull/29721) since it seems that this bug should be fixed  for systems without rstat flush funcion. Please take a look:-) Thanks:-)

xxhdx1985126 commented:
@ukernel Hi, thanks again for your spending so much time(probably a whole day) reviewing my code:-)
I just can't thank you enough on this:-) Really apprieciate it:-)

xxhdx1985126 commented:
@ukernel 
Hi, I've refactored this PR according all the above discussions, and modified the rstat flush test case to do the test more aggresively. Please take a look:-)

By the way, this PR also includes the commit in PR:[29721](https://github.com/ceph/ceph/pull/29721), since it needs that commit to run correctly.

Thanks:-)

xxhdx1985126 commented:
jenkins retest this please

xxhdx1985126 commented:
jenkins test submodule

xxhdx1985126 commented:
jenkins test submodules

xxhdx1985126 commented:
jenkins test sign-off-by

xxhdx1985126 commented:
jenkins retest this, please

xxhdx1985126 commented:
```
hudson.plugins.git.GitException: Failed to fetch from https://github.com/ceph/ceph.git
	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:894)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1161)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1192)
	at hudson.scm.SCM.checkout(SCM.java:504)
	at hudson.model.AbstractProject.checkout(AbstractProject.java:1208)
	at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:574)
	at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:499)
	at hudson.model.Run.execute(Run.java:1818)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:429)
```

The build keep failing due to the above error. Does this mean there's something wrong with the building environment?

xxhdx1985126 commented:
jenkins retest this please

xxhdx1985126 commented:
jenkins retest this please

xxhdx1985126 commented:
jenkins retest this please

xxhdx1985126 commented:
jenkins retest this please

ukernel commented on src/mds/Locker.cc:
```diff
@@ -4786,12 +4831,70 @@ void Locker::scatter_tick()
+        dout(20) << __func__ << " " << (mdr ? mdr->reqid : metareqid_t())
+          << " need to nudge scatterlock: " << *lock << " of " << *lock->get_parent() << dendl;
+
+        scatter_nudge(lock, new C_Locker_RstatFlush(this, mdr, lock));
```
still allocate a C_Locker_RstatFlush for each lock?  I think it's better to use a MDSGatherBuilder and only allocate one C_Locker_RstatFlush. In my option,  CInode::need_to_nudge is not a cheap function

xxhdx1985126 replied:
@ukernel Thanks for your review:-) I'm trying to modify the code as you suggested and make CInode::need_to_nudge cheap:-)

ukernel commented on src/mds/Locker.cc:
```diff
@@ -4786,12 +4831,70 @@ void Locker::scatter_tick()
+    if (mdr->in[0]->is_projected() && mdr->client_request && !mdr->already_replied) {
+      mdr->in[0]->add_waiter(CInode::WAIT_POPPROJECTED,
+	  new C_Locker_RespondToRequest(this, mdr));
+      mdr->already_replied = true;
```
you'd better to use mdlog->wait_for_safe()

xxhdx1985126 replied:
@ukernel Hi, I think we have to make sure clients get the response after the projected inode is poped, if we use "mdlog->wait_for_safe()" there's still a possibility that the C_Locker_RespondToRequest is invoked before the Context that pops the projected inode, in which case there exists a chance that the clients' later getattr retrieves an out-dated rstat. Am I right? Thanks:-)

ukernel commented on src/mds/MDCache.cc:
```diff
@@ -13050,3 +13096,167 @@ void MDCache::handle_mdsmap(const MDSMap &mdsmap) {
+    dir->get_inode()->make_path(req->srcdnpath);
+    req->op_stamp = mdr->get_op_stamp();
+    mds->send_message_mds(req, dir->get_dir_auth().first);
+    internal_mdr->more()->queried_dir = dir;
```
why not just set origin base of rstat flush. If we do this, there is no reason to create multiple slave requests.

xxhdx1985126 replied:
@ukernel I'll try to follow your way:-)

ukernel commented:
On Tue, Sep 3, 2019 at 3:12 PM Xuehan Xu <notifications@github.com> wrote:

> *@xxhdx1985126* commented on this pull request.
> ------------------------------
>
> In src/mds/Locker.cc
> <https://github.com/ceph/ceph/pull/29629#discussion_r320119622>:
>
> > +  // If no locks is of the current mdr's interest, all rstats are flushed successfully
> +  if (mdr && !nudged) {
> +    dout(20) << __func__ << " " << mdr->reqid
> +      << " rstat flush finished!" << dendl;
> +
> +    // If the inode is projected and we reply clients a success ack,
> +    // there exists chances that later getattrs, who comes before
> +    // the projected inode poped and retrieves the non-projected
> +    // inode attrs, get an out-dated rstat.
> +    //
> +    // So we wait until the current projected_inode is poped to ack
> +    // the clients.
> +    if (mdr->in[0]->is_projected() && mdr->client_request && !mdr->already_replied) {
> +      mdr->in[0]->add_waiter(CInode::WAIT_POPPROJECTED,
> +	  new C_Locker_RespondToRequest(this, mdr));
> +      mdr->already_replied = true;
>
> @ukernel <https://github.com/ukernel> Hi, I think we have to make sure
> clients get the response after the projected inode is poped, if we use
> "mdlog->wait_for_safe()" there's still a possibility that the
> C_Locker_RespondToRequest is invoked before the Context that pops the
> projected inode, in which case there exists a chance that the clients'
> later getattr retrieves an out-dated rstat. Am I right? Thanks:-)
>
>
contexts is executed in their add order. If you add
C_Locker_RespondToRequest after projecting inode. it will get executed
after popping inode


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceph/ceph/pull/29629?email_source=notifications&email_token=AAVFTL5VGHOKSSQZVA6U37TQHYE5JA5CNFSM4ILHCFO2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDNSPJA#discussion_r320119622>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAVFTL5JAUEQZCHFOW34TW3QHYE5JANCNFSM4ILHCFOQ>
> .
>

xxhdx1985126 commented:
@ukernel Got it:-) Will change that part of the code:-)

xxhdx1985126 commented:
@ukernel Hi, I've modified the code as you suggested:-)

xxhdx1985126 commented:
jenkins test make check

xxhdx1985126 commented:
@ukernel Hi, I've just modified some commits' messages so that it can describe the implemetation more precisely, please take a look, thanks:-)

xxhdx1985126 commented:
@ukernel @batrick there was some mistake in the commit messages, I just corrected them. please take a look, thanks:-)

xxhdx1985126 commented:
jenkins test make check

xxhdx1985126 commented:
jenkins retest this, please

batrick commented:
@xxhdx1985126 Zheng is out for a few weeks. He'll review again as soon as he can.

xxhdx1985126 commented:
@batrick Got it:-) Thanks:-)

xxhdx1985126 commented:
@ukernel  Hi, I've just rearranged some of the code and added some more comments to make to code more comprehensive. Please take a look, thanks:-)

ukernel commented:
PR is still too complex to understand the whole picture. I started to implement a simple version

xxhdx1985126 commented:
@ukernel Hi, There is a total description of the rstat flush procedure in commit "mds: flush all rstat to root":-)

ukernel commented:
See https://github.com/ceph/ceph/pull/32126.

xxhdx1985126 commented:
please unmark

xxhdx1985126 commented:
please unmark

mchangir commented:
> PR is still too complex to understand the whole picture. I started to implement a simple version

@ukernel if possible, could you describe the idea for the simpler patch ?
